mne/preprocessing/asr_utils.py

```diff
@@ -0,0 +1,509 @@
+            ne = int(np.fix(F[i + 1] * npt)) - 1
+
+        j = np.arange(nb, ne + 1)
+        if ne == nb:
+            inc = 0
+        else:
+            inc = (j - nb) / (ne - nb)
+
+        Ht[0][nb:ne + 1] = np.array(inc * M[i + 1] + (1 - inc) * M[i])
+        nb = ne + 1
+
+    Ht = np.concatenate((Ht, Ht[0][-2:0:-1]), axis=None)
+    n = Ht.size
+    n2 = np.fix((n + 1) / 2)
+    nb = order
+    nr = 4 * order
+    nt = np.arange(0, nr)
+
+    # compute correlation function of magnitude squared response
+    R = np.real(np.fft.ifft(Ht * Ht))
+    R = R[0:nr] * (0.54 + 0.46 * np.cos(np.pi * nt / (nr - 1)))   # pick NR correlations  # noqa
+
+    # Form window to be used in extracting the right "wing" of two-sided
+    # covariance sequence
+    Rwindow = np.concatenate(
+        (1 / 2, np.ones((1, int(n2 - 1))), np.zeros((1, int(n - n2)))),
+        axis=None)
+    A = polystab(denf(R, order))  # compute denominator
+
+    # compute additive decomposition
+    Qh = numf(np.concatenate((R[0] / 2, R[1:nr]), axis=None), A, order)
+
+    # compute impulse response
+    _, Ss = 2 * np.real(signal.freqz(Qh, A, worN=n, whole=True))
+
+    hh = np.fft.ifft(
+        np.exp(np.fft.fft(Rwindow * np.fft.ifft(np.log(Ss, dtype=np.complex))))  # noqa
+    )
+    B = np.real(numf(hh[0:nr], A, nb))
+
+    return B, A
+
+
+def yulewalk_filter(X, sfreq, zi=None, ab=None, axis=-1):
+    """Yulewalk filter.
+
+    Parameters
+    ----------
+    X : array, shape = (n_channels, n_samples)
+        Data to filter.
+    sfreq : float
+        Sampling frequency.
+    zi : array, shape=(n_channels, filter_order)
+        Initial conditions.
+    a, b : 2-tuple | None
+        Coefficients of an IIR filter that is used to shape the spectrum of
+        the signal when calculating artifact statistics. The output signal
+        does not go through this filter. This is an optional way to tune the
+        sensitivity of the algorithm to each frequency component of the
+        signal. The default filter is less sensitive at alpha and beta
+        frequencies and more sensitive at delta (blinks) and gamma (muscle)
+        frequencies.
+    axis : int
+        Axis to filter on (default=-1, corresponding to samples).
+
+    Returns
+    -------
+    out : array
+        Filtered data.
+    zf :  array, shape=(n_channels, filter_order)
+        Output filter state.
+    """
+    # Set default IIR filter coefficients
+    if ab is None:
+        F = np.array([0, 2, 3, 13, 16, 40, np.minimum(
+            80.0, (sfreq / 2.0) - 1.0), sfreq / 2.0]) * 2.0 / sfreq
+        M = np.array([3, 0.75, 0.33, 0.33, 1, 1, 3, 3])
+        B, A = yulewalk(8, F, M)
+    else:
+        A, B = ab
+
+    # apply the signal shaping filter and initialize the IIR filter state
+    if zi is None:
+        out = signal.lfilter(B, A, X, axis=axis)
+        zf = None
+    else:
+        out, zf = signal.lfilter(B, A, X, zi=zi, axis=axis)
+
+    return out, zf
+
+
+def ma_filter(N, X, Zi):
```

Comment:
I think this should be done with https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.lfilter.html#scipy.signal.lfilter